llvm/lib/Analysis/HashRecognize.cpp

```diff
@@ -0,0 +1,683 @@
+getRecurrences(BasicBlock *LoopLatch, const PHINode *IndVar, const Loop &L) {
+  std::optional<RecurrenceInfo> SimpleRecurrence, ConditionalRecurrence;
+  for (PHINode &P : LoopLatch->phis()) {
+    if (&P == IndVar)
+      continue;
+    if (!P.getType()->isIntegerTy()) {
+      LLVM_DEBUG(dbgs() << "HashRecognize: Non-integral PHI found\n");
+      return {};
+    }
+
+    BinaryOperator *BO;
+    Value *Start, *Step;
+    const APInt *GenPoly;
+    if (!SimpleRecurrence && matchSimpleRecurrence(&P, BO, Start, Step)) {
+      SimpleRecurrence = {&P, BO, Start, Step};
+    } else if (!ConditionalRecurrence &&
+               matchConditionalRecurrence(&P, BO, Start, Step, L, GenPoly,
+                                          Instruction::BinaryOps::Xor)) {
+      ConditionalRecurrence = {&P, BO, Start, Step, *GenPoly};
+    } else {
+      LLVM_DEBUG(dbgs() << "HashRecognize: Stray PHI found: " << P << "\n");
+      return {};
+    }
+  }
+  return {SimpleRecurrence, ConditionalRecurrence};
+}
+
+PolynomialInfo::PolynomialInfo(unsigned TripCount, const Value *LHS,
+                               const APInt &RHS, const Value *ComputedValue,
+                               bool ByteOrderSwapped, const Value *LHSAux)
+    : TripCount(TripCount), LHS(LHS), RHS(RHS), ComputedValue(ComputedValue),
+      ByteOrderSwapped(ByteOrderSwapped), LHSAux(LHSAux) {}
+
+/// In big-endian case, checks that bottom N bits against CheckFn, and that the
+/// rest are unknown. In little-endian case, checks that the top N bits against
+/// CheckFn, and that the rest are unknown.
+static bool checkExtractBits(const KnownBits &Known, unsigned N,
+                             function_ref<bool(const KnownBits &)> CheckFn,
+                             bool ByteOrderSwapped) {
+  unsigned BitPos = ByteOrderSwapped ? 0 : Known.getBitWidth() - N;
+  unsigned SwappedBitPos = ByteOrderSwapped ? N : 0;
+
+  // Check that the entire thing is a constant.
+  if (N == Known.getBitWidth())
+    return CheckFn(Known.extractBits(N, 0));
+
+  // Check that the {top, bottom} N bits are not unknown and that the {bottom,
+  // top} N bits are known.
+  return CheckFn(Known.extractBits(N, BitPos)) &&
+         Known.extractBits(Known.getBitWidth() - N, SwappedBitPos).isUnknown();
+}
+
+/// Generate a lookup table of 256 entries by interleaving the generating
+/// polynomial. The optimization technique of table-lookup for CRC is also
+/// called the Sarwate algorithm.
+CRCTable HashRecognize::genSarwateTable(const APInt &GenPoly,
+                                        bool ByteOrderSwapped) const {
+  unsigned BW = GenPoly.getBitWidth();
+  unsigned MSB = 1 << (BW - 1);
+  CRCTable Table;
+  Table[0] = APInt::getZero(BW);
+
+  if (ByteOrderSwapped) {
+    APInt CRCInit(BW, 1);
+    for (unsigned I = 1; I < 256; I <<= 1) {
+      CRCInit = CRCInit.shl(1) ^
+                ((CRCInit & MSB).isZero() ? APInt::getZero(BW) : GenPoly);
```

Comment:
```suggestion
                (CRCInit.isSignBitSet() ? GenPoly : APInt::getZero(BW));
```
and remove the `MSB` variable above.